cms/cms_menus.py

```diff
@@ -209,10 +212,9 @@ def get_nodes(self, request):
             fallbacks = languages
 
         pages = (
-            pages
-            .filter(pagecontent_set__language__in=languages)
```

Comment:
Good catch! The `pages` queryset gets turned into a list when `get_visible_nodes` is called. Let me have a look at `get_visible_nodes` to return a queryset instead of a list. A first glance, it gives the impression there is room for further performance improvement. Maybe we can save another query?

Reply:
@marksweb Now, I've looked into that. `pages` is pushed as an iterable into `prefetch_related`.  So the prefetch is done manually there.

Keeping it as a queryset, is I hoped would improve the database accesses, therefore will not do so. Also, the permissions cache is filled, which - if `get_visible_nodes` needed to return a qs - would not be possible and might lead to performance penalties later. Finally, the permission rules are quite complicated. Turning them into queries is not immediately obvious. 

I would stay with the current structure.